reorganize configuration syntax intro (#849)
grafana/alloy: docs/sources/concepts/configuration-syntax/_index.md

# {{% param "PRODUCT_NAME" %}} configuration syntax

{{< param "FULL_PRODUCT_NAME" >}} dynamically configures and connects components with the {{< param "PRODUCT_NAME" >}} configuration syntax. Overall, {{< param "FULL_PRODUCT_NAME" >}} handles the collection, transformation, and delivery of telemetry data, and so each component in the configuration handles one of those tasks, or specifies how data flows (for example from collection to transformation), and so how the components are bound together.

{{< figure src="/media/docs/alloy/flow-diagram-small-alloy.png" alt="Alloy flow diagram" >}}

Here is a simple example of a fully worked configuration, showing the basic concepts, and how an Alloy config comes together into a pipeline.

```alloy
// Collection: mount a local directory with a certain path spec
local.file_match "applogs" {
    path_targets = [{"__path__" = "/tmp/app-logs/app.log"}]

> }

Reviewer: `local.file_match` is a weird one, as it fits into the same bucket as the `discovery` components. It's not collecting any data, but performs discovery to enrich other components.

Do you think this distinction is useful at the introduction level? If so, maybe it could be a fourth category of components so the categories become:

1. Data enrichment (local.file_match, discovery.kubernetes, etc.) 
2. Collection 
3. Transformation
4. Delivery 

I'm ok with either; I'll defer to your judgment for what makes the most sense for high-level introductory content.

Author: I borrowed this from a worked example my team did, I don't have anything wrapped up in its specifics.

If we can keep the example technically correct while making it shorter & simpler, that's a win.  We specifically don't need to teach anything advanced here, it's only there to demonstrate the "pipeline/flow" concept with a simple example anyone can follow (report logs)

// Collection: Take the file match as input, and scrape those mounted log files
loki.source.file "local_files" {
    targets    = local.file_match.applogs.targets

    // This specifies which component should process the logs next, the "link in the chain"
    forward_to = [loki.process.add_new_label.receiver]
}

// Transformation: pull some data out of the log message, and turn it into a label
loki.process "add_new_label" {
    stage.logfmt {
        mapping = {
            "extracted_level" = "level",
        }
    }

    // Add the value of "extracted_level" from the extracted map as a "level" label
    stage.labels {
        values = {
            "level" = "extracted_level",
        }
    }

    // The next link in the chain is the local_loki "receiver" (receives the telemetry)
    forward_to = [loki.write.local_loki.receiver]
}

// Anything that comes into this component gets written to the loki remote API
loki.write "local_loki" {
    endpoint {
        url = "http://loki:3100/loki/api/v1/push"
    }
}
```

> The {{< param "PRODUCT_NAME" >}} syntax aims to reduce errors in configuration files by making configurations easier to read and write.
> {{< param "PRODUCT_NAME" >}} configurations use blocks that can be easily copied and pasted from the documentation to help you get started as quickly as possible.

Reviewer 2: ```suggestion
The {{< param "PRODUCT_NAME" >}} syntax aims to reduce errors in configuration files by making configurations easier to read and write.
The {{< param "PRODUCT_NAME" >}} syntax uses blocks, attributes, and expressions.
The blocks can be copied and pasted from the documentation to help you get started as quickly as possible.
```

The sentence about blocks, attributes, and expressions was out of place in the flow. I moved it up here, and suggest deleting it from line 64. We might need to look into how lines 61 and 62 fit into the flow of ideas here too...

The {{< param "PRODUCT_NAME" >}} syntax is declarative, so ordering components, blocks, and attributes does not matter.
The relationship between components determines the order of operations in the pipeline.

> The {{< param "PRODUCT_NAME" >}} syntax uses blocks, attributes, and expressions, which are now explained, with examples.
> 

Reviewer 2: ```suggestion

```

> ## Blocks

Reviewer: Are we concerned about any potential confusion now that blocks (which are composed of attributes) are explained before attributes?

Reviewer: If this is a concern, but the ordering is still preferred, maybe we can add some kind of teaser to explain attributes, something like: 

> You use _Blocks_ to configure components and groups of settings called attributes.

Reviewer 2: I think this suggestion is more clear and a good replacement for line 68.

You form a reference to a component's exports by merging the component's name (for example, `local.file`),
label (for example, `password_file`), and export name (for example, `content`), delimited by a period.

## Configuration syntax design goals

> {{< param "PRODUCT_NAME" >}} is designed with the following requirements in mind:

Reviewer 2: ```suggestion
{{< param "PRODUCT_NAME" >}} is:
```

We could simplify this down to a more direct statement

> * _Fast_: The configuration language must be fast so the component controller can quickly evaluate changes.
> * _Simple_: The configuration language must be easy to read and write to minimize the learning curve.
> * _Debuggable_: The configuration language must give detailed information when there's a mistake in the configuration file.

Reviewer 2: ```suggestion
* _Fast_: The configuration language is fast, so the component controller can quickly evaluate changes.
* _Simple_: The configuration language is easy to read and write to minimize the learning curve.
* _Easy to debug_: The configuration language gives detailed information when there's a mistake in the configuration file.
```

Simplify to be more direct

The {{< param "PRODUCT_NAME" >}} configuration syntax is a distinct language with custom syntax and features, such as first-class functions.

* Blocks are a group of related settings and usually represent creating a component.
  Blocks have a name that consists of zero or more identifiers separated by `.`, an optional user label, and a body containing attributes and nested blocks.
* Attributes appear within blocks and assign a value to a name.
* Expressions represent a value, either literally or by referencing and combining other values.

>   You use expressions to compute a value for an attribute.

Reviewer: Likely not in scope for this PR, but I wonder if this content adds anything new and if it makes more sense to merge it into other headers. 

WDYT about the long-term value of this section @moxious?

Reviewer 2: I was thinking the same. These points feel like a repetition or a rephrase of previous info... and it might be better to find a way to rephrase and fit it into the content around lines 59-65... somewhere in that area.  Might help to commit some of the other changes and do a `make docs` so this can all be read in the rendered output.

Author: Noting the goals of the design for a configuration syntax is, in my view - of primary interest to the people who want to hack on the syntax or contribute to core alloy.  So it's pretty advanced material.  That's why I chose to locate it at the bottom.  It wasn't in the scope of this PR to try to remove something, but being frank - if you're a beginning user you probably won't read this, because just knowing how to use the config syntax is enough.

If I were to move this content, I would move it into a section that's about "how to contribute to alloy" or "why we did what we did".  Very useful information to gate what kind of contributions would be welcome.
